linkerd/app/src/env.rs

```diff
@@ -547,10 +547,23 @@ pub fn parse_config<S: Strings>(strings: &S) -> Result<super::Config, EnvError>
                 Some(addr) => {
                     // If the inbound is proxy is configured to discover policies, then load the set
                     // of all known inbound ports to be discovered during initialization.
-                    let mut ports = match parse(strings, ENV_INBOUND_PORTS, parse_port_set)? {
-                        Some(ports) => ports,
+                    let mut ports = match strings.get(ENV_INBOUND_PORTS)? {
+                        Some(s) => {
+                            if s.is_empty() {
+                                warn!("No inbound ports specified via {}", ENV_INBOUND_PORTS,);
+                                Default::default()
+                            } else {
+                                parse_port_set(&s).map_err(|parse_error| {
+                                    error!(
+                                        "{}={:?} is not valid: {:?}",
+                                        ENV_INBOUND_PORTS, s, parse_error
+                                    );
+                                    EnvError::InvalidEnvVar
+                                })?
+                            }
+                        }
                         None => {
-                            error!("No inbound ports specified via {}", ENV_INBOUND_PORTS,);
+                            warn!("No inbound ports specified via {}", ENV_INBOUND_PORTS,);
```

Comment:
I'm open to suggestions on simplifying this. I originally tried something similar to

```rust
Some("") | None => {
    warn!("No inbound ports specified via {}", ENV_INBOUND_PORTS,);
    Default::default()
}
Some(s) => {
    ...
}
```

But we cannot match on `Some("")` unfortunately. Therefore, we have the duplicate handling of issuing the `warn!` and setting `ports = Default::default()`.

Reply:
I believe the reason we cannot match `Some` here is because the value is an `Option<&String>` rather than an `Option<&str>`, if it was an `&str`, `Some(''")` should work. I think if we call `as_deref()` on the option before matching it, the empty string pattern should work. See this playground as an example: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6fd8540f90081c6360ac5c10de336ae8

Reply:
Ah that's great!